stages/oscap.remediation: Properly utilize offline capabilities

evgenyz opened the pull request:
![Screenshot from 2023-10-02 14-21-43](https://github.com/osbuild/osbuild/assets/429134/0f2f7e6f-d941-454b-9ecd-e21ec3e5ba04)
The scanner will properly react to the chroot environment. Also, there are some optimizations to logs and results.

The stage is now divided into four steps:
- Initial scan
- Generation of fix script
- Execution of the script
- Cleanup

It is necessary to properly set up scanner probes into offline mode, which allows to gracefully ignore absence of `/proc` and other stuff where it is possible. 

Since the scanner does not support direct remediation with `--remediate` in offline mode we must go through intermediary steps.

On the bright side it will save us second-after-remediation scan (which is not entirely correct anyway). To get the real evaluation of the system we must execute the scanner in a live environment. 

<s>For the same reason I've got rid of HTML report. The ARF results file is way more informative if used with `oscap-report` tool.</s> Nope. No backward-incompatible changes.

Fixes: #1360

evgenyz commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +61,64 @@ SCHEMA = """
     # optional vars
     xccdf_id = config.get("xccdf_id")
-    data_dir = options.get("data_dir")
+    data_dir = options.get("data_dir", DATA_DIR)
```
Since we need a place to work with remediation script, the data_dir is now mandatory and defaults to `/root`. Which would be a convenient place to store the remediation artifacts for analyses.

evgenyz commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -128,15 +129,82 @@ def main(tree, options):
+    res = subprocess.run(cmd, encoding="utf8", stdout=sys.stderr, check=False,
+                         env=dict(os.environ,
+                             OSCAP_PROBE_ROOT="",
+                             OSCAP_CONTAINER_VARS="container=bwrap-osbuild"))
```
With this variable set scanner would respect `osbuild` platform of the CaC content and won't try to enable FIPS and other stuff, which would just ruin the image if executed in way that is meant for live systems.

evgenyz commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -128,15 +129,82 @@ def main(tree, options):
+        # run with command in chroot so full path ok
+        cmd.extend(["--results-arf", f"{data_dir}/{arf_results}"])
+
+    cmd.extend(["--results", f"{data_dir}/{xccdf_results}"])
```
This result file is now a mandatory artifact as it is a source for fix generation step later on.

evgenyz commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -128,15 +129,82 @@ def main(tree, options):
+
+    print("Pack the result files: ", repr(cmd))
+    res = subprocess.run(cmd, encoding="utf8", stdout=sys.stderr, check=False)
 
```
The default footprint in `data_dir` (`/root` if omitted) is:
- oscap_eval_xccdf_results.xml.xz (up to 10 Mb in the worst case)
- oscap_remediation.bash (about 1 Mb).

kingsleyzissou replied:
@achilleas-k, is there a better way we could handle compressing the output?

thozza replied:
I would say that this is OK, but the approach has a few downsides...

The stage previously didn't depend on `xz` being available in the build environment. Therefore old manifest which don't have it installed in the build pipeline, would start failing if they used this stage. For the sake of backwards compatibility, I would suggest to make the compression optional (based on the availability of `xz` or based on an explicit schema property).

evgenyz replied:
How exactly should I check for presence of `xz` in a canonical way if there is any?

thozza replied:
There is no standard way that we use AFAIK, so from the top of my head I would say:

1. Check that `/usr/bin/xz` exists before using it.
2. Gracefully handle the error from `subprocess.run()`.

However, thinking about this a bit more, I think that it may be better to make the compression an explicit option in the stage schema, instead of changing the stage behavior based on the build environment. This would make the stage behavior more predictable. The old behavior (not compressing anything) can be preserved by default, while the stage can fail in case one enables the compression in stage options and `xz` is not available.

evgenyz replied:
Well there was no 150 MB file stored in /root in old behavior. That's not nice for sure, I'd like this file to be compressed by default. Also, it is impossible to turn this option on from a blueprint so it won't be very convenient.

kingsleyzissou replied:
> it is impossible to turn this option on from a blueprint

We can handle that internally in composer, so there would be no change to the blueprints

evgenyz replied:
Alright, explicit options it is.

kingsleyzissou commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -128,15 +129,82 @@ def main(tree, options):
+    cmd = [
+        "/usr/sbin/chroot", tree,
+        "/usr/bin/oscap",
+        "xccdf", "generate"
```
I am wondering if we should maybe at this point split this into 3 stages then?
- scan
- generate
- remediate

osbuild-composer would take care of passing through these parameters and we could handle the `data-dir` and results from there.

@supakeen & @thozza, what are your thoughts? I'm assuming this might be a backwards compatibility issue?

evgenyz replied:
It might not be necessary in the future. We just can't do this change in the scanner right away. Also, it'd require change in OpenSCAP to generate new oscap entry(ies) for the blueprint.

thozza replied:
@kingsleyzissou splitting the stage into three pieces is not a problem from backwards compatibility PoV, as long as the original stage keeps working as it did before. This would probably mean introducing three new stages.

Nevertheless, my opinion on whether to split the stage or not is that it would make sense only if each of those new stages would be useful on its own. If one would need to always use all three of those stages in combination to get some useful result, then it makes more sense to keep the implementation within a single stage.

kingsleyzissou replied:
Yeah, very unlikely that they would be used in isolation from each other, so we can keep them as is then :)

kingsleyzissou commented:
@evgenyz thanks for the PR, really impressed by how much this affects the compliance score!

evgenyz commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -128,15 +129,82 @@ def main(tree, options):
+    # does not have access to /proc
+    res = subprocess.run(cmd, encoding="utf8", stdout=sys.stderr, check=False,
+                         env=dict(os.environ,
+                             OSCAP_PROBE_ROOT="",
```
Done.

evgenyz commented:
> @evgenyz thanks for the PR, really impressed by how much this affects the compliance score!

Well, we also did our homework in OpenSCAP and CaC. But yeah, proper offline mode is important for offline systems :)

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -128,15 +130,88 @@ def main(tree, options):
+
+    # pack result files, it could save dozens of megabytes
+    cmd = [
+        "/usr/sbin/chroot", tree,
+        "/usr/bin/xz"
```
Please note the dependency of this stage on `xz` and `chroot` (should have been noted already :confused: ) in the stage doc text at the top of the file.

kingsleyzissou replied:
:grimacing: the `chroot` one was my bad

thozza reviewed (commented):
Thanks for these improvements.

I added a few comments inline, but the most important problem I see is the backward incompatibility of your changed. We treat stage schema as an API, thus old schema potentially (and in reality) used by existing osbuild manifests must keep working after your modifications.

evgenyz commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +61,65 @@ SCHEMA = """
         "type": "string",
-        "description": "Filename for saving the html report",
-        "default": "eval_remediate_report.html"
+        "description": "Filename for storing the XCCDF Results file"
```
Sure. The `html_report` is back.

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +63,76 @@ SCHEMA = """
         "type": "string",
         "description": "The tailoring id"
       },
-      "xml_results": {
+      "arf_results": {
```
AFAICT, this means that the `xml_results` stage options key is no longer accepted by the schema, which is backward incompatible. If it is a synonym for `xccdf_results`, you need to use `anyOf` in the schema.

evgenyz replied:
Yep, that's the idea. Hope I got it right now.

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -6,10 +6,12 @@ The OpenSCAP scanner can be run on the image tree and the remediation can be car
-The stage generates an html report and xml results file both saved to the `/openscap_data`
-directory.
+The stage generates an html report and xml results file both saved to the `data_dir`
+directory. It defaults to `/root` if not configured.
```
Please note the default in the schema below. In addition, is there some good reason to change the default from `/openscap_data` ?

evgenyz replied:
I believe it never actually was `/openscap_data` by default, the doc lied. Also, it feels like littering.

thozza replied:
Yeah, the path was not explicitly used by the stage before. Since I'm not really familiar with the `oscap` tool, I simply assumed that the `/openscap_data` is the path used by the tool in case no path is specified. If this is not the case, then the change looks good (I mean, it looked good in any case, I was mostly curious).

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +63,76 @@ SCHEMA = """
-XML_RESULTS = "eval_remediate_results.xml"
-HTML_REPORT = "eval_remediate_report.html"
+DATA_DIR = "/root"
+XCCDF_RESULTS = "oscap_eval_xccdf_results.xml"
```
please note this default in the stage schema

evgenyz replied:
Done.

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +63,76 @@ SCHEMA = """
-    html_report = config.get("html_report", HTML_REPORT)
+    arf_results = config.get("arf_results")
+    xccdf_results = config.get("xccdf_results", XCCDF_RESULTS)
+    html_report = config.get("html_report", "report.html")
```
please note this default in the stage schema.

evgenyz replied:
Done.

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +63,76 @@ SCHEMA = """
+    arf_results = config.get("arf_results")
+    xccdf_results = config.get("xccdf_results", XCCDF_RESULTS)
+    html_report = config.get("html_report", "report.html")
+    verbose_log = config.get("verbose_log", "log")
+    verbose_level = config.get("verbose_level", "INFO" if verbose_log is not None else None)
```
please note the default values  for both properties in the stage schema.

in addition, `verbose_log` can't ever be `None`, because if no value was provided, it will be `"log"`.

evgenyz replied:
Fixed.

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +63,76 @@ SCHEMA = """
+    if verbose_level is not None:
+        cmd.extend(["--verbose", verbose_level])
+
+    if verbose_log is not None:
```
`verbose_log` can't ever be `None`, since the default value is `"log"` if not provided to the stage in its options.

evgenyz replied:
Fixed.

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +63,76 @@ SCHEMA = """
-        # run with command in chroot so full path ok
-        cmd.extend(["--results", f"{data_dir}/{xml_results}"])
-        cmd.extend(["--report", f"{data_dir}/{html_report}"])
+    if verbose_level is not None:
```
`verbose_level` can't ever be `None`, since it is always set to some value and by default to `"INFO"`, because `verbose_log` is always set.

evgenyz replied:
Removed debug hardcode, sorry about it slipping through.

thozza reviewed (commented):
Thanks for your changes. I added some more comments. Unfortunately, the stage schema changes are still backward incompatible AFAICT.

thozza reviewed (commented):
Thanks for all the changes. The PR looks good, with the one exception which is on me, sorry...

One last thing - I think that it would be good to add a variant of the stage unit test https://github.com/osbuild/osbuild/tree/main/test/data/stages/oscap.remediation which would test the new options. Basically making a copy and using different stage options.

thozza commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -25,6 +27,7 @@ SCHEMA = """
 "properties": {
   "data_dir": {
     "type": "string",
+    "default: "/root",
```
You are missing one double-quote at the end of `default`

evgenyz replied:
Oh!

evgenyz commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -61,63 +64,83 @@ SCHEMA = """
-        "default": "eval_remediate_results.xml"
-      },
-      "arf_results": {
+      "anyOf": [
```
Yeah, makes sense. No worries, fixed.

evgenyz replied:
I got rid of `oneOf` in the end. Proper implementation makes the scheme over-complicated with no real benefit.

evgenyz commented:
> One last thing - I think that it would be good to add a variant of the stage unit test https://github.com/osbuild/osbuild/tree/main/test/data/stages/oscap.remediation which would test the new options. Basically making a copy and using different stage options.

Good idea, stay tuned.

kingsleyzissou commented on stages/org.osbuild.oscap.remediation:
```diff
@@ -128,15 +151,131 @@ def main(tree, options):
+    cmd.extend(["fix", "--fix-type", "bash",
+                "--output", f"{data_dir}/{REMEDIATION_SCRIPT}"])
+
+    cmd.append(f"{data_dir}/{xccdf_results}")
```
@evgenyz going through this again, I have a qq about tailoring. Will this step need the tailoring file too if it's specified? Or are the generated results enough for this?

evgenyz replied:
Results should be enough (tailored-out rules are not processed and therefore won't be fixed).

kingsleyzissou replied:
Apologies for re-opening this. I rebased some tailoring changes on this and notice that unfortunately this does break :/

With `composer` we overwrite the `profile_id` to the new tailoring `profile_id` (it just appends "_osbuild_tailoring" to the base profile). So this step breaks because it's looking for the tailoring profile.

I guess we could either: 
 - check for the suffix "_osbuild_tailoring" and save those into `base_profile_id` and `tailoring_profile_id`
 - add a `base_profile_id` in as a schema option, which is probably what I should have done in the first place.
 
The only problem I see with the second option is it breaks backwards compatibility. Maybe we could do both? Wdyt @thozza?

kingsleyzissou replied:
@evgenyz is there any way we can use `oscap` to get the base `profile_id` from the tailoring file?

evgenyz replied:
So, with `oscap info tail.xml` you can get the base profile id. It's in a human-readable form, but should be stable.

evgenyz replied:
So there is no need to pass the tailoring profile id. I'll change my PR.

kingsleyzissou replied:
Amazing, that will be the simplest. Thanks so much

evgenyz replied:
Well, actually it turned out that `generate fix` unlike other generators actually needs the same tailored file. Sorry for misguiding you. Nonetheless, I've updated the PR to reflect that. You should not change anything in the autotailor step, it does what it is supposed to do.

evgenyz commented:
> > One last thing - I think that it would be good to add a variant of the stage unit test https://github.com/osbuild/osbuild/tree/main/test/data/stages/oscap.remediation which would test the new options. Basically making a copy and using different stage options.
> 
> Good idea, stay tuned.

So, I've added a test. Fixed existing one. And removed `oneOf` completely because it actually requires a lot of shenanigans in the schema. I think we should be just fine having these 2 parameters: the arf_results will just take precedence.

thozza reviewed (commented):
Thanks for all the changes and new test.

I'm on the fence WRT there being two properties in the schema options, which are synonyms, can be specified at the same time and one takes precedence over the other. It feels weird, especially when one can specify different values in each of them. The behavior is not obvious and not documented.

Could you maybe elaborate more on the "schema shenanigans" that made you to take this approach? IOW, was something not working or was it too difficult to make things work?

The PR technically looks OK from my PoV, but not approving yet, since I want to have some discussion and consensus on the stage schema changes related to the synonymous properties.

evgenyz commented:
> Could you maybe elaborate more on the "schema shenanigans" that made you to take this approach? IOW, was something not working or was it too difficult to make things work?

There is no way of simply wrapping two keys (not values) in a `oneOf` directive. These must be objects like in 
https://github.com/osbuild/osbuild/blob/main/stages/org.osbuild.udev.rules#L178. Maybe I'm missing something
here but it looks pretty complicated for a property that basically controls debug output (if you set any of two or all of them you will get a result file in the end, no changes in scan/remediation logic).

But I can switch back to `oneOf` and fully implement the required scheme if you think it really worth it.

> The PR technically looks OK from my PoV, but not approving yet, since I want to have some discussion and consensus on the stage schema changes related to the synonymous properties.

Sure.

thozza reviewed (approved):
From what I can tell, the PR looks good. Thanks for your effort and patience! Nevertheless, @kingsleyzissou should review and approve before merging, since he knows the most about OSCAP.

kingsleyzissou commented:
I've verified it with tailoring and it works, so besides the linting issues it looks good to me.

evgenyz commented:
> I've verified it with tailoring and it works, so besides the linting issues it looks good to me.

Sorry, in the end, in order to shut it I used a technique inspired by `org.osbuild.grub2`. The code is just too *branchy*.

thozza commented:
@evgenyz silencing pylint using a directive in the code would be also fine from my PoV...

kingsleyzissou reviewed (approved):
Great, thanks again for your work and your patience on this :smile: :rocket:

thozza reviewed (changes requested):
Nice work.

However, I have one request - please make the fix of the existing test part of the commit which made it so that this is needed (presumably the first commit). Alternatively, you can just squash both commits to a single one.

My reasoning is to ensure that unit tests pass just fine also for the first commit alone, which is desired for potential bisecting.

evgenyz commented:
> However, I have one request - please make the fix of the existing test part of the commit which made it so that this is needed (presumably the first commit). Alternatively, you can just squash both commits to a single one.

Done, everything is in a single commit.

thozza reviewed (approved):
Awesome, thanks a lot!
